lnrpc/lightning.proto

```diff
@@ -200,6 +200,16 @@ service Lightning {
     */
     rpc OpenChannel (OpenChannelRequest) returns (stream OpenStatusUpdate);
 
+    /* lncli: `batchopenchannel`
+    BatchOpenChannel attempts to open multiple single-funded channels in a
+    single transaction in an atomic way. This means either all channel open
+    requests succeed at once or all attempts are aborted if any of them fail.
+    This is the safer variant of using PSBTs to manually fund a batch of
+    channels through the OpenChannel RPC.
+    */
+    rpc BatchOpenChannel (BatchOpenChannelRequest)
```

Comment:
What is your thought on this earlier comment?

https://github.com/lightningnetwork/lnd/pull/5356#issuecomment-857399439

Reply:
I don't think this really applies to this RPC. The operation is really atomic, so it either goes through (and the channels will be pending) or it doesn't (and everything is cleaned up and canceled), no matter what happens to the RPC request.
Or do you see a case where things aren't fully reverted with the current code?

Reply:
The case that I see is that you launch this call and then crash, lose connection, etc. How does the caller find out whether the funding process is in progress or not? In the happy flow, the channels eventually show up as pending. But should the caller wait for a safe amount of time (how long?) before retrying the batch call? I am not sure what happens if funding fails, but perhaps nothing is seen other than messages in the log?

Reply:
Right, I agree that the case you describe isn't easy to handle. But I don't think there is anything in this particular PR that we need to add to fix the issue in https://github.com/lightningnetwork/lnd/issues/5329. What we need is a new RPC that returns any channel that is in the state where only a (non-persisted) reservation exists for it, including its pending channel ID. And possibly also add an RPC to cancel a reservation by its ID.

Reply:
Yes, my point was just to think through how this v2 channel open rpc can be used if resumability is added later. To avoid needing to create v3.

Your suggestion to create a `listchannels` for pre-pending - would that be sufficient? Suppose two applications connected to lnd would both do the exact same `batchopen` call. One of them fails, but that application will think it succeeded based on the `listchannels` output that shows the actions of the other `batchopen` call.

In #5329 I suggested a call identifier so that it will always be possible to track progress later. But that is something that can be added later on and doesn't require any changes right now.

Reply:
I think a call identifier is a good idea. In fact, we already have the pending channel ID that is the unique identifier of a channel while it is in the reservation state. We use that for the current v1 PSBT funding flow. So I'm going to add that as an optional field to the batch channel so an API user can already set it today instead of it being randomly generated (but will need the RPC to list pre-pending channels to query their state).

Reply:
Is it right that even without that list rpc, this addition already makes `batchopen` safe to call multiple times? So if you call it a second time with the same IDs (for example after disconnect), the call fails (with a structured error)?

Reply:
Hmm, good question. I haven't tried that but it's possible that we check if a channel with the pending ID already exists.

Reply:
I hope that at the minimum nothing bad happens in this case.

Reply:
Okay, I tried this with the example you mentioned [here](https://github.com/lightningnetwork/lnd/issues/5329#issuecomment-848917291). If you use the same `pending_chan_id`, you get an error as expected `Unknown desc = batch funding failed: error batch opening channel, initial negotiation failed: pendingChannelID(3f55d436d4f7d24aeffb1b976647380f22ebf9e74390e8c76dcff9fea0093b7a) already present`.

BUT: You need to check the pending channels first! If the channel went through and the channel is actually pending, then using the same `pending_chan_id` won't result in an error (because we never store that ID anywhere other than in the reservation, which is removed in `handleFundingCreated` and `handleFundingSigned`).

I think this isn't perfect but a bit better than what you currently get with the normal channel open flow. So I won't change anything more in this PR and leave further optimizations for the PR that fixes #5329.

Reply:
Yes that is much better already! Only a tiny gap left between checking pending channels and re-trying the channel open.

So if the channel is already pending and you open a new channel with the same id, you will just get two channels without any other undesired side effects?

Reply:
> So if the channel is already pending and you open a new channel with the same id, you will just get two channels without any other undesired side effects?

Yes. The pending channel ID is only used as the main in-memory key for the reservation, it is not used for anything later on. The "real" channel ID that is used for persisting the channel is derived from the channel point.